### ISSUE TITLE
Seek most-overlapping types when relating object types in unions

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#26450]
From @rkirov at https://github.com/DefinitelyTyped/DefinitelyTyped/issues/28044#issue-349690446

```ts
interface A {
    a: string;
    b: number
};

interface Other {
    o: number;
}

declare function f(x: A | Other): any;

export let x = { a: '', b: '' };
export let y = f(x);
```

**Expected**

```
Argument of type '{ a: string; b: string; }' is not assignable to parameter of type 'Other | A'.
  Type '{ a: string; b: string; }' is not assignable to type 'A'.
    Types of property 'b' are incompatible.
      Type 'string' is not assignable to type 'number'.
```

**Actual**

```
Argument of type '{ a: string; b: string; }' is not assignable to parameter of type 'A | Other'.
  Type '{ a: string; b: string; }' is not assignable to type 'Other'.
    Property 'o' is missing in type '{ a: string; b: string; }'.
```

**Comment by bowenni** [comment]
I'm not sure if this is a valid ask. Would it be possible to use the most-overlapping declaration in error message for overloads too?

```typescript
declare function f(a: string, b: string, c: string): void;
declare function f(a: () => string, b: () => string, c: () => string): void;
f('1', '1', {});
```
Actual
`ERROR(3,3): : Argument of type '"1"' is not assignable to parameter of type '() => string'.`
Expected
`ERROR(3,13): : Argument of type '{}' is not assignable to parameter of type 'string'.`

**Comment by DanielRosenwasser** [comment]
@bowenni I think that is a UX improvement we want to see, but I'd rather we file a different issue there. Overload resolution is pretty complicated and would warrant its own work item.

**Comment by CyrusNajmabadi** [comment]
A vote on this over from Pulumi land.  We have extensive use of union types, and errors here can be painful :)

**Comment by DanielRosenwasser** [comment]
Doesn't "dive in" to the objects like @joeduffy and @CyrusNajmabadi were hoping for, but a fix is up on #27087.
